File: depth_first_search/1315_sum_of_nodes_with_even_valued_grandparent/1315_sum_of_nodes_with_even_valued_grandparent.py
import collections

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def sumEvenGrandparent(self, root):
        grandfa = father = None
        res = 0
        deque = collections.deque()
        deque.append((root, father, grandfa))
        while deque:
            node, father, grandfa = deque.popleft()
            if node:
                if grandfa and grandfa % 2 == 0:
                    res += node.val
                deque.append((node.left, node, father))
                deque.append((node.right, node, father))

File: depth_first_search/1315_sum_of_nodes_with_even_valued_grandparent/1315_sum_of_nodes_with_even_valued_grandparent.go
package main 

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

func sumEvenGrandparent(root *TreeNode) int {
	if root == nil {
		return 0
	}

	sum := 0

	if root.Val % 2 == 0 {
		if root.Right != nil {
			if root.Right.Left != nil {
				sum += root.Right.Left.Val
			}
			if root.Right.Right != nil {
				sum += root.Right.Right.Val
			}
		}

		if root.Left != nil {
			if root.Left.Left != nil {
				sum += root.Left.Left.Val
			}
			if root.Left.Right != nil {
				sum += root.Left.Right.Val
			}
		}
	}

	sum += sumEvenGrandparent(root.Left)
	sum += sumEvenGrandparent(root.Right)

	return sum
}